### ISSUE TITLE
PDF export file size increases hugely from v3.18 to v3.20

### ISSUE THREAD

**Issue by TonyM-J** [qgis/QGIS#44656]
### What is the bug or the crash?

I have a standard layout that I export as a .pdf. Up to QGIS 3.18 the pdf file size was typically about 2.7MB, since moving to 3.20.1 it has increased to about 17MB for exactly the same layout and data. I can see nothing in the release notes about changes to the export function.

Reinstalling 3.18 resulted in the file size returning to about 2.7MB. 

[EMMA Layout-multi_3_18.pdf](https://github.com/qgis/QGIS/files/6967983/EMMA.Layout-multi_3_18.pdf)
[EMMA Layout-multi_3_20.pdf](https://github.com/qgis/QGIS/files/6967986/EMMA.Layout-multi_3_20.pdf)

### Steps to reproduce the issue

Try exporting an identical layout as a pdf from version 3.18 and version 3.20

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.18.3-Zürich | QGIS code revision | 735cc85be9
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.1.4 | Running against GDAL/OGR | 3.1.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Windows 10 (10.0)
Active python plugins | emma_creator; mmqgis; pluginbuilder3; plugin_reloader; TomBio; db_manager; processing

</body></html>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.20.1-Odense | QGIS code revision | 1c3c5cd6
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.5
GDAL/OGR version | 3.3.1
PROJ version | 8.1.0
EPSG Registry database version | v10.018 (2021-04-02)
GEOS version | 3.9.1-CAPI-1.14.2
SQLite version | 3.35.2
PDAL version | 2.3.0
PostgreSQL client version | 13.0
SpatiaLite version | 5.0.1
QWT version | 6.1.3
QScintilla2 version | 2.11.5
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins | emma_creatormmqgispluginbuilder3plugin_reloaderTomBiodb_managerprocessing

</body></html>

**Comment by gioman** [comment]
On a very simple project/layout with just one single vector layer the size is more than double on 3.20 (on Windows)

Anyway I tested on master on Linux and the size is exactly the same. Seems to be that it could be caused by some underlying library, on Windows there are more recent dependencies compared to my Ubuntu 20.04 machine.

**Comment by TonyM-J** [comment]
That helpfully narrows the problem down to a change in the Windows library that is used to prepare the pdf. I have no idea how to identify this or revert to a previous version.

**Comment by agiudiceandrea** [comment]
AFAIK the Qt library it's responsible for creating the PDF file.
Current Qt version in old (v1/msi) OSGeo4W is 5.11, in new (v2/msi) is 5.15.
Moreover, with https://github.com/qgis/QGIS/pull/37594 was introduced a new option to control PDF image compression method when exporting layouts to PDF (it's only available for QGIS >= 3.16.0 and Qt >= 5.13).
Is the "Image compression" option set to "Lossy (JPEG)" or to "Lossless"?

**Comment by gioman** [comment]
>  Is the "Image compression" option set to "Lossy (JPEG)" or to "Lossless"?

@agiudiceandrea in both cases is set to the same value, lossy.




>  Current Qt version in old (v1/msi) OSGeo4W is 5.11, in new (v2/msi) is 5.15.

So possibly a regression upstream.

**Comment by agiudiceandrea** [comment]
> I've just checked using v3.20.1 (Qt 5.15.2) on Win 10. Using my setup, exporting a layout with the 'lossy (JPEG)' option (default) selected gives a pdf size of 17,8 MB and exporting the same layout with 'lossless' selected gives exactly the same pdf size.

@TonyM-J I didn't look carefully at your pdf files...
It seems they don't contain "images" (images, raster layers or rasterized vector layers), so the "Image compression" option has no effect in your case.

> By the way, is this option exposed in the Python API? If so, what's the option name? There doesn't appear to be a relevant option in PdfExportSettings()

I think you just need to set a flag.
See: https://qgis.org/api/structQgsLayoutExporter_1_1PdfExportSettings.html#a8ef337cee21992b46b4c5db06d2e2f2a https://qgis.org/api/classQgsLayoutRenderContext.html#aae823feceb47451b481ce3be51e37456
```
pdf_settings = QgsLayoutExporter.PdfExportSettings()
pdf_settings.flags = QgsLayoutRenderContext.FlagLosslessImageRendering
```

**Comment by TonyM-J** [comment]
Thank you @agiudiceandrea; you are quite correct that my pdfs don't contain images/rasterized layers, so image compression will have no effect and the output pdf file size should be the same in both cases (as it is). That means that the reason for the  8x increase in pdf size between 3.18 and 3.20 lies elsewhere - presumably somewhere in the Qt library that does the pdf rendering. 
Thanks also for the info about setting flags - that might be useful in some contexts.

**Comment by gioman** [comment]
> how is set the stroke width of your layers symbology? Is it set to 0 / Hairline?

@agiudiceandrea difference in size is noticeable also by just using a point layer with a default symbology.

**Comment by agiudiceandrea** [comment]
> @agiudiceandrea difference in size is noticeable also by just using a point layer with a default symbology.

A point layer with a default symbology has a "Hairline" stroke width.

**Comment by TonyM-J** [comment]
@agiudiceandrea, @gioman - That seems to be the problem. With all 4 layers set to a stroke width of 0.2pt the output pdf size is 2.1MB. Resetting the stroke width for all layers to 'hairline' increases the pdf size to 17MB. So this bug, introduced between QT 5.11 and 5.15, is the problem.

**Comment by GisUsers** [comment]
The issue is also present in QGIS 3.16.10 (v2) on Win 10.

**Comment by agiudiceandrea** [comment]
> The issue is also present in QGIS 3.16.10 (v2) on Win 10.

Yes, of course. The issue depends on the Qt library version that has a bug, not the QGIS version.
All the QGIS versions (3.16, 3.20, 3.21 dev, 3.18) installed using the new v2 OSGeo4W installers suffer from the reported issue because the Qt version installed by the new v2 OSGeo4W installlers is the version 5.15.2, which has the bug.

**Comment by agiudiceandrea** [comment]
See also the issue report: https://github.com/qgis/QGIS/issues/44876
So it seem the bigger PDF file could be due to the fact that the "Hairlines" are actually incorrectly created as "polygons" in the PDF files by Qt 5.15.2 on Windows (OSGeo4W v2) instead of "Hairlines".

**Comment by nyalldawson** [comment]
This can be closed here -- it's being addressed upstream and there's nothing we can do to fix it on the qgis side.

**Comment by agiudiceandrea** [comment]
The underlying Qt bug https://bugreports.qt.io/browse/QTBUG-86094 has been reportedly fixed.

**Comment by GisUsers** [comment]
It seems the underlying bug has been fixed in Qt version 5.15.7
Is it possible to predict when it will make its way to QGIS?

**Comment by agiudiceandrea** [comment]
Hi @jef-n, updating Qt5 in OSGeo4W to a version >= 5.15.7, if possible, will fix this issue and the https://github.com/qgis/QGIS/issues/44876 issue on Windows.
